src/nomad_simulations/schema_packages/basis_set.py

```diff
@@ -0,0 +1,642 @@
+                '`APWPlaneWaveBasisSet.cutoff_fractional` already defined. Will not overwrite.'
+            )  #! extend implementation
+            return
+        elif self.cutoff_energy is None or mt_r_min is None:
+            logger.warning(
+                '`APWPlaneWaveBasisSet.cutoff_energy` and `APWPlaneWaveBasisSet.radius` must both be defined. Aborting normalization step.'
+            )
+            return
+        self.cutoff_fractional = self.cutoff_radius.to(
+            f'1 / {reference_unit}'
+        ) * mt_r_min.to(reference_unit)
+
+
+class AtomCenteredFunction(ArchiveSection):
+    """
+    Specifies a single function (term) in an atom-centered basis set.
+    """
+
+    pass
+
+    # TODO: design system for writing basis functions like gaussian or slater orbitals
+
+
+class AtomCenteredBasisSet(BasisSet):
+    """
+    Defines an atom-centered basis set.
+    """
+
+    functional_composition = SubSection(
+        sub_section=AtomCenteredFunction.m_def, repeats=True
+    )  # TODO change name
+
+    def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+        super().normalize(archive, logger)
+        # self.name = self.m_def.name
+        # TODO: set name based on basis functions
+        # ? use naming BSE
+
+
+class APWBaseOrbital(ArchiveSection):
+    """
+    Abstract base section for (S)(L)APW and local orbital component wavefunctions.
+    It helps defining the interface with `APWLChannel`.
+    """
+
+    n_terms = Quantity(
+        type=np.int32,
+        description="""
+        Number of terms in the local orbital.
+        """,
+    )
+
+    energy_parameter = Quantity(
+        type=np.float64,
+        shape=['n_terms'],
+        unit='joule',
+        description="""
+        Reference energy parameter for the augmented plane wave (APW) basis set.
+        Is used to set the energy parameter for each state.
+        """,
+    )  # TODO: add approximation formula from energy parameter n
+
+    energy_parameter_n = Quantity(
+        type=np.int32,
+        shape=['n_terms'],
+        description="""
+        Reference number of radial nodes for the augmented plane wave (APW) basis set.
+        This is used to derive the `energy_parameter`.
+        """,
+    )
+
+    energy_status = Quantity(
+        type=MEnum('fixed', 'pre-optimization', 'post-optimization'),
+        default='post-optimization',
+        description="""
+        Allow the code to optimize the initial energy parameter.
+        """,
+    )
+
+    differential_order = Quantity(
+        type=np.int32,
+        shape=['n_terms'],
+        description="""
+        Derivative order of the radial wavefunction term.
+        """,
+    )  # TODO: add check non-negative # ? to remove
+
+    def _get_open_quantities(self) -> set[str]:
+        """Extract the open quantities of the `APWBaseOrbital`."""
+        return {
+            k for k, v in self.m_def.all_quantities.items() if self.m_get(v) is not None
```

Comment:
you cannot do `getattr(self, k)` directly and forget about the values?

Reply:
Honestly, idk what the proper NOMAD section interface is: it changes all the time...
Your suggestion is appropriate for vanilla classes, yes.